test/vmtester/vmtester.cpp

```diff
@@ -28,29 +26,131 @@ evmc_instance* get_vm_instance()
     return vm.get();
 }
 
+class cli_parser
+{
+public:
+    const char* const app_name = nullptr;
+    const char* const app_version = nullptr;
+    const int max_num_args = 0;
+
+    std::vector<std::string> args;
+
+    bool help = false;
+    bool version = false;
+
+    std::string output;
+    std::string errors;
+
+    cli_parser(const char* app_name, const char* app_version, int max_num_args) noexcept
+      : app_name{app_name}, app_version{app_version}, max_num_args{max_num_args}
+    {
+        args.reserve(static_cast<size_t>(max_num_args));
+    }
+
+    int parse(int argc, char* argv[])
+    {
+        int num_args = 0;
+        for (int i = 1; i < argc; ++i)
+        {
+            auto arg = std::string{argv[i]};
+
+            auto x = arg.find_first_not_of('-');
+            if (x == 0)  // Argument.
+            {
+                ++num_args;
+                if (num_args > max_num_args)
+                {
+                    errors += "Unexpected argument \"" + arg + "\"\n";
+                    return -1;
+                }
+                args.emplace_back(std::move(arg));
+                continue;
+            }
+            else if (x <= 2)
+            {
+                arg.erase(0, x);
+
+                if (arg == "version")
+                {
+                    version = true;
+                    continue;
+                }
+                if (arg == "help")
+                {
+                    help = true;
+                    continue;
+                }
+            }
+
+            errors += "Unknown option \"" + std::string{argv[i]} + "\"\n";
+            return -1;
+        }
+
+        output += app_name + std::string{" "} + app_version + "\n";
+
+        if (help)
+        {
+            output += std::string{"Usage: "} + argv[0] + " " + "MODULE" + "\n";
+            return 0;
+        }
+
+        if (version)
+            return 0;
+
+        if (!version && num_args < max_num_args)
+        {
+            errors += "The MODULE argument is required.\n";
+            errors += "Run with --help for more information.\n";
+            return -1;
+        }
+
+        return 0;
+    }
+};
+
 int main(int argc, char* argv[])
 {
     try
     {
-        std::string vm_path;
-
-        CLI::App app{"EVMC VM Tester"};
-        app.add_option("vm", vm_path, "Path to the VM shared library to be tested");
         testing::InitGoogleTest(&argc, argv);
-        CLI11_PARSE(app, argc, argv);
 
-        std::cout << "Testing " << vm_path << "\n";
+        auto cli = cli_parser{"EVMC VM Tester", PROJECT_VERSION, 1};
+
+        if (auto error_code = cli.parse(argc, argv))
+        {
+            std::cerr << cli.errors;
+            return error_code;
+        }
+
+        if (cli.help)
+        {
+            std::cout << "\n" << cli.output;
```

Comment:
Not sure why you need the newline, but you could add it in the parsing code like `if (help) { output = "\n" + output + ...; }`

Reply:
New line is nice because GTest outputs the message for --help first.

But I reworked this again, by taking `cout` and `cerr` as arguments.